File: leetcode3/염혜정/914. X of a Kind in a Deck of Cards.java
// O(nlogn)

import java.util.Arrays;

class Solution {
    public boolean hasGroupsSizeX(int[] deck) {
        Arrays.sort(deck);

        int g = 0;             
        int count = 1;          
        for (int i = 1; i <= deck.length; i++) {
            if (i < deck.length && deck[i] == deck[i - 1]) {
                count++;        
            } else {
                g = gcd(g, count);  
                count = 1;
            }
        }
        return g >= 2;
    }

    private int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }
}

File: leetcode3/최원준/914. X of a Kind in a Deck of Cards.py
from collections import defaultdict
class Solution:
    def hasGroupsSizeX(self, deck: List[int]) -> bool:
        counter = defaultdict(int)
        for d in deck:
            counter[d] += 1
        
        count = counter[deck[0]]
        for _, freq in counter.items():
            if freq == 1:
                return False
            count = gcd(count, freq)

        return True if count != 1 else False
    
